theseus/core/robust_cost_function.py

```diff
@@ -107,15 +113,25 @@ def weighted_jacobians_error(self) -> Tuple[List[torch.Tensor], torch.Tensor]:
             weighted_jacobians,
             weighted_error,
         ) = self.cost_function.weighted_jacobians_error()
+        if self.flatten_dims:
+            weighted_error = weighted_error.reshape(-1, 1)
+            for i, wj in enumerate(weighted_jacobians):
+                weighted_jacobians[i] = wj.diagonal(dim1=1, dim2=2).reshape(-1, 1, 1)
         squared_norm = torch.sum(weighted_error**2, dim=1, keepdim=True)
         rescale = (
             self.loss.linearize(squared_norm, self.log_loss_radius.tensor)
             + RobustCostFunction._EPS
         ).sqrt()
 
-        return [
+        rescaled_jacobians = [
             rescale.view(-1, 1, 1) * jacobian for jacobian in weighted_jacobians
```

Comment:
Following my previous comments, this will be something like `torch.einsum("ni, nij->nij", rescale.view(-1, self.dims()), jacobians)`

Reply:
Similar comment as above.

Reply:
With .diagonal, it presumes that the Jacobian matrix is a square matrix. And for a jacobian, it would be dE/dX, which determines its shape is (dim(E), dim(x)). My question is, will it be necessarily equal for dim(E) and dim(X)?

Reply:
Good observation! I forgot to mention that this implementation assumes that we want to model N robust cost functions of dim=1, each with $k$ variables of dof=1, with a single robust cost function of dimension N and $k$ variables of dof=N. Basically, this assumes you want to use a single `RobustCost` object to model a problem of the form

$$J =  \sum_{i=1:N} \rho(f_i(x_i, y_i, ...)^2)$$

Supporting cost function of dim > 1 and other combinations of variables dof gets messier, and I'm not sure it's worth the effort.

Note that this feature is mostly for programming convenience. Creating individual cost functions objects doesn't hurt performance, because `TheseusLayer` by defaults vectorizes cost functions with the same pattern so they are evaluated in a single call.

Reply:
But it seems it's not feasible for the case like in tutorial 4, motion model (First figure). In that case, the Error is the combination of 2 vectors, which could be usual and convenient in many case. Therefore, in my original code, I obtain the dimension of the input X through the Jacobian matrix. Because the error itself only contains the batch size and the dimension of the error. And I believe based on the chain rule the dRobustLoss/dOriLoss should be able to be multiplied by the original Jocobian directly.
![image](https://user-images.githubusercontent.com/52436595/235299755-e4e80c6e-81fe-4707-baf8-3bdfeb868a31.png)
![image](https://user-images.githubusercontent.com/52436595/235300203-f98b05e8-5ea6-4235-b79b-7a3ecee9fb5b.png)
![image](https://user-images.githubusercontent.com/52436595/235300235-8d770965-f7cd-48f4-9d98-9886673795b1.png)

Reply:
A more appropriate example is the pose graph optimization case where each error term is dim=6. You can see that we model each error term with its [own separate `RobustCost`](https://github.com/facebookresearch/theseus/blob/main/examples/pose_graph/pose_graph_synthetic.py#L161), rather than having one large `RobustCost` that computes all error terms at once. There is no need for a `flatten_dims` argument or anything similar. The loss penalizes only those edges in the graph that might be the result of very noise relative pose measurements, it doesn't matter that the cost has dim > 1.

Reply:
Of course, we can always construct cost functions that are not supported by our implementation. For example, if you combine measurements from two dim > 1 sources in a single cost function, then this wouldn't do what you want. But I would argue that in that case rewriting as separate cost terms would be cleaner and less error prone.

Reply:
Yes. You are right. Currently, I do not see an optimization problem that would result in where dim(E) != dim(X) or cannot be decoupled into multiple cost functions that fulfill dim(E) = dim(X). We can further discuss it if one encounters those problems. Thank you! I have learned a lot from your project!

Reply:
No problem, and thank you for all your useful feedback!

Reply:
@xphsean12 I have to apologize, because turns out I was wrong in some of the above! We definitely want to support the case were dim(error) != dim(X). For example, in the regression problem, dim(error) = N_points, but dim(X) = 1 both for A and B. In the above I was mixing up optim vars with aux vars. 

Luckily, the fix is still pretty simple, and I have rewritten the test to check for the case above, using the same regression example. You can look at the last few commits if you are curious. 

The code still assumes that `flatten_dims` acts as if it was dim(err) independent costs of dim=1.

Reply:
Thanks for your feedback! By the way, for the radius,  (maybe) I found a new issue. Since I do not have very concrete evidence, I am not going to raise an issue so far. In my previous commit, I change the radius = radius.exp() into .square(). @fantaosha mention that exp() is better for differentiating. However, the easiest approach should be just let radius = radius. And we just need to let the user input an radius >= 0.

The main difference is that (1) based on my observation, .exp() is much slower than radius = radius when I try to train the Huber radius (2) training an value, which will be further pass to the exp() function may be very difficult. Because when changing the trained value from 0 to 5, the actual radius being used is changed from 1 to 148. The exponential growth may cause training an accurate radius difficult. Because the radius will be enlarged by an exponential function. I am not sure whether my second concern is correct, just kind of thinking.